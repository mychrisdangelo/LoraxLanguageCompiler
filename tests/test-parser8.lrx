/*
 * Author: Chris D'Angelo
 * Lorax Parser Kitchen Sink
 * Testing valid parseable file
 */

// parsing requires global variables must be declared first
int a;
tree <float>b(3);
float c;
char d;
string e;

int inc (int x) {
	return x + 1;
}

char capitalize_letter_a (char a) {
	if (a == 'a') {
		return 'A';
	}
	return '0';
}

int change_first_child_letter_to_p(tree <char>r(2)) {
	r%(1-1)@ = 'w'; // for fun
	// r%1-1@ = 'w'; // this is acceptable syntax but not semantics
	r%0@ = 'p';
	return 0;
}

int change_letter_to_q(tree <char>n(2)) {
	n@ = 'q';
	return 0;
}

int print_for_me_please(string s)
{
	print(s + "\n");
}

int capitalize_all_of_me(string s)
{
	string tmp;
	tmp = s;
	while (tmp%0 != null) {
		if (tmp@ < 'z') {
			// lowercase
			tmp@ = tmp@ + 'A' - 'a';
		}
		tmp = tmp%0;
	}
	return 0;
}

int main() {
	// parsing requires function locals must be declared first
	tree <float>g(3);
	tree <char>k(2);
	int l;
	char m;
	bool s;
	bool t;
	string v;
	tree <char>z(2);
	int y;
	l = 2;
	a = 4;

	while (l < a) {
		inc(l);
		break;
	}
	y = -1;
	print(a mod 3);
	g = 1.1[2.1, 2.2[2.21, 2.22, 2.23], 2.3[2.31, 2.32]];

	k = 'z'['x', 'y'['b', 'a']];
	print(capitalize_letter_a(k%1%1));

	change_letter_to_q(k%1%1);
	print(k%1%1@); // should print 0 and tree should be 'z'['x', 'q'['b', 'a']];
	change_first_child_letter_to_p(k);
	print(k); // should print 0 and tree should be 'z'['p', 'q'['b', 'a']];

	print(t = (!s || false && true));

	print_for_me_please("hello");

	v = "hello";
	capitalize_all_of_me(v);
	print(v);

	z = k + 'm'['n', 'o']; // will give 'a' a child 'm' (which itself has two children)

	z%1--; // pop the second (ref: 1) child off of z
	z = 'm'['n', 'o']--; // will nullify this tree
	z = 'm'['n', 'o']%0--; // will pop the 'n' from the tree

	for (l = 0; l < 42; l = l + 1) {
		print(l);
	}

	a = b = c;
	z%0%3@ = 4;

	a = z%0;

	t%0 = t2; // t and t2 are both trees. t2 is now being assigned the first child of t (accessed by %0)
	t%0 = t3%0; // similar to above but now t's first child is t3's first child.
	t%0@ = 3; // @ dereferences the value in that node. Now the t's first child's value is 3
	t%z@ = 4;
	normal_int = t%0@; // now we're assigning a normal int var the value from inside t's first child value
	t@ = 4; // this is assigning the root nodes value as 4
	t%0%1@ = 5; // this is assigning t's first child's, second child node value to 5
	t3 = t%0%1--; // this is popping t's first child's second child node from the tree t and returns t to assign to t3
	t%toyfunc()%3-- = t2;
}