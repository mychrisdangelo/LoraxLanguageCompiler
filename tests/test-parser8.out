tree <char>e(1);
char d;
float c;
tree <float>b(3);
int a;
int main()
{
tree <float>g(3);
tree <char>k(2);
int l;
char m;
bool s;
bool t;
tree <char>v(1);
tree <char>z(2);
int y;
l = 2;
a = 4;
while (l < a) {
inc(l);
break;}
y = -1;
print(a mod 3);
g = 1.1[2.1, 2.2[2.21, 2.22, 2.23], 2.3[2.31, 2.32]];
k = 'z'['x', 'y'['b', 'a']];
print(capitalize_letter_a(k % 1 % 1));
change_letter_to_q(k % 1 % 1);
print(k % 1 % 1@);
change_first_child_letter_to_p(k);
print(k);
print(t = !s || false && true);
print_for_me_please("hello");
v = "hello";
capitalize_all_of_me(v);
print(v);
z = k + 'm'['n', 'o'];
z % 1--;
z = 'm'['n', 'o']--;
z = 'm'['n', 'o'] % 0--;
for (l = 0 ; l < 42 ; l = l + 1) {
print(l);
}
a = b = c;
z % 0 % 3@ = 4;
a = z % 0;
t % 0 = t2;
t % 0 = t3 % 0;
t % 0@ = 3;
t % z@ = 4;
normal_int = t % 0@;
t@ = 4;
t % 0 % 1@ = 5;
t3 = t % 0 % 1--;
t % toyfunc() % 3-- = t2;
}

int capitalize_all_of_me(tree <char>s(1))
{
tree <char>tmp(1);
tmp = s;
while (tmp % 0 != null) {
if (tmp@ < 'z')
{
tmp@ = tmp@ + 'A' - 'a';
}
tmp = tmp % 0;
}
return 0;
}

int print_for_me_please(tree <char>s(1))
{
print(s + "\n");
}

int change_letter_to_q(tree <char>n(2))
{
n@ = 'q';
return 0;
}

int change_first_child_letter_to_p(tree <char>r(2))
{
r % 1 - 1@ = 'w';
r % 0@ = 'p';
return 0;
}

char capitalize_letter_a(char a)
{
if (a == 'a')
{
return 'A';
}
return '0';
}

int inc(int x)
{
return x + 1;
}